### DISCUSSION TITLE
template structure

### DISCUSSION THREAD

**Discussion by younesStrittmatter** [AutoResearch/autora#370] [Ideas]
Instead of a template for each contribution (experimentalist, theorist, runner) ... we could have one template that includes all of in the appropriate folder (src/theorist/your-theorist), (src/experimentalist/sampler/your-sampler), ... and explain the approriate folder to use for each contribution in the readme? (would make it easier to maintain since we don't have to update each template individually and maybe even has an educational effect to show contributors how the whole project is organized)

**Comment by younesStrittmatter** [comment]
Is there any reason why we shouldn't do this? We can even add a README in each folder for specifics (Having too many templates will just be painful to maintain since every little change has to be distributed to all of the templates

**Reply by musslick** [reply]
I do see the appeal. Although it does raise the bar for contributions and user errors (e.g. if they don't delete the irrelevant folders after use). E.g., the user could copy a template, adds an experimentalist but keeps all the irrelevant folders and demo files for the example theorist in their package.

**Reply by musslick** [reply]
Also pinging @hollandjg here; this seems like a decision worth making sooner rather than later. I'm certainly worried about the template duplication, as it has already caused some inconvenience. As an alternative, is there a way where we can "mirror" a base template?

**Reply by younesStrittmatter** [reply]
Seems like another category for a golden raspberry to me:) First team member to publish an empty theorist, experimentalist, or runner under the namespace .../example

**Reply by musslick** [reply]
Yeah, I am worried about that too. That said, I very much like the idea of having a single place where we make adaptations to the template... That said, the individual templates are somewhat different in that they have different tests and documentation requirements...

**Reply by hollandjg** [reply]
I'd much prefer a single template. I don't think there's a way to mirror a template and add something on top, so I think we'll just need good documentation.

**Reply by hollandjg** [reply]
... and someone will definitely publish something wrong at some point. It's a good experience to have.

**Reply by musslick** [reply]
I guess it's settled. Let's go with a single template for now and add proper documentation.

**Reply by gtdang** [reply]
Perhaps we can look into [Cookiecutter](https://cookiecutter.readthedocs.io/en/stable/README.html).  It's a project templating package that essentially clones template repositories but then can rename folders and file attributes based on variables defined in a yaml file. Its advanced usage seems to have hooks that can conditionally remove files and folders. So potentially have a way to remove irrelevant folders and files depending on what type of contribution it will be?

**Reply by musslick** [reply]
That sounds great! Does it work through GitHub Actions? It might be easiest for the contributor to clone the repo and then run a Github Action on it that removes irrelevant folders....

**Reply by gtdang** [reply]
No, it's actually a CLI app. So you install it with homebrew or pip. Then you give the command `cookiecutter <RepoURL>` and then it will clone the template repo and ask you a series of questions. These questions could be package name,  metadata to populate in the pyproject.toml, etc. You can also have conditional questions like: Theorist [y/n], Experimentalist[y/n], etc. Those conditionals would delete files/folders from the template as needed.

**Reply by gtdang** [reply]
I'll try to work up a demo for next week's maintenance meeting.

**Reply by gtdang** [reply]
I've got a demo version of the Cookiecutter working with the autora template. On cloning the repo it will ask questions like if this is for a theorist or experimentalist, then can perform functions to trim and edit the template based on your contribution type. It will also fill in information to any files (notably the pyproject.toml) based on user-supplied info so the user doesn't have hunt for all the fields they need to change. 

This also allows us to just maintain one template repository while not over-loading contributors with directories and files they will need to manually delete. 

I can show it this afternoon at our maintenance meeting.
